File: jas_ocaml/lib/interpreter/point_buffers.ml
(** Thread-local-equivalent (ref-cell) point buffers.

    Tools that accumulate coordinates during a drag (Lasso, Pencil)
    use named buffers here. The module-local assoc list plays the
    role of Rust's thread_local! / Swift's module ref. *)

let buffers : (string * (float * float) list) list ref = ref []

let clear (name : string) : unit =
  buffers := (name, []) :: List.filter (fun (k, _) -> k <> name) !buffers

let push (name : string) (x : float) (y : float) : unit =
  let existing = match List.assoc_opt name !buffers with
    | Some l -> l | None -> []
  in
  let updated = existing @ [(x, y)] in
  buffers := (name, updated) :: List.filter (fun (k, _) -> k <> name) !buffers

let length (name : string) : int =
  match List.assoc_opt name !buffers with
  | Some l -> List.length l
  | None -> 0

let points (name : string) : (float * float) list =
  match List.assoc_opt name !buffers with
  | Some l -> l
  | None -> []

File: workspace_interpreter/point_buffers.py
"""Thread-local-equivalent point buffers for drag-accumulating tools.

Tools like Lasso and Pencil push (x, y) coordinates here during a
drag. Module-local dict plays the role of Rust's thread_local! and
Swift's module ref — Qt/GTK run single-threaded so a plain dict
suffices.
"""

from __future__ import annotations


_buffers: dict[str, list[tuple[float, float]]] = {}


def clear(name: str) -> None:
    """Reset the named buffer to empty."""
    _buffers[name] = []


def push(name: str, x: float, y: float) -> None:
    """Append (x, y) to the named buffer."""
    _buffers.setdefault(name, []).append((x, y))


def length(name: str) -> int:
    return len(_buffers.get(name, []))


def points(name: str) -> list[tuple[float, float]]:
    """Return a shallow copy of the buffer's points."""
    return list(_buffers.get(name, []))
